substrate/client/network/sync/src/state_strategy.rs

```diff
@@ -0,0 +1,330 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! State sync strategy.
+
+use crate::{
+	schema::v1::StateResponse,
+	state::{ImportResult, StateSync},
+	types::{BadPeer, OpaqueStateRequest, OpaqueStateResponse, SyncState, SyncStatus},
+};
+use libp2p::PeerId;
+use log::{debug, error, info, trace};
+use sc_client_api::ProofProvider;
+use sc_consensus::{BlockImportError, BlockImportStatus, IncomingBlock};
+use sp_consensus::BlockOrigin;
+use sp_runtime::{
+	traits::{Block as BlockT, NumberFor},
+	Justifications, SaturatedConversion,
+};
+use std::{collections::HashMap, sync::Arc};
+
+/// Log target for this file.
+const LOG_TARGET: &'static str = "sync";
+
+mod rep {
+	use sc_network::ReputationChange as Rep;
+
+	/// Peer response data does not have requested bits.
+	pub const BAD_RESPONSE: Rep = Rep::new(-(1 << 12), "Incomplete response");
+
+	/// Reputation change for peers which send us a known bad state.
+	pub const BAD_STATE: Rep = Rep::new(-(1 << 29), "Bad state");
+}
+
+/// Action that should be performed on [`StateStrategy`]'s behalf.
+pub enum StateStrategyAction<B: BlockT> {
+	/// Send state request to peer.
+	SendStateRequest { peer_id: PeerId, request: OpaqueStateRequest },
```

Comment:
Overall sounds good, but one thing that limits us is the `BlockDownloader` trait introduced in https://github.com/paritytech/polkadot-sdk/pull/1524. When we replace all requests by `RequestType` and `payload: Vec<u8>`, I don't think there will be an easy way to plug custom block downloading implementations.

One option that might work is to make  `BlockDownloader` run on top of requests. I.e., it generates opaque vector of bytes to send over network, and handles the opaque bytes received as a response. But this will essentially limit it to substrate request-response protocols.

@rahulksnv will this work for you, or you are using some custom protocol not based on substrate request-response protocols?

Reply:
Currently we are using `BlockDownloader` to send custom messages instead of what Substrate sends by default to implement something very similar to [Compact Blocks (BIP 0152)](https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki) and with plans to implement [Graphene](https://cryptoeconomics.cs.umass.edu/graphene/) at some point in the future.

> One option that might work is to make BlockDownloader run on top of requests. I.e., it generates opaque vector of bytes to send over network, and handles the opaque bytes received as a response

IIRC this is basically how it works already.

@rahulksnv can you confirm what I just wrote?

Reply:
If you are fetching transactions from "mempool", downloading the missing ones, do I understand correctly that `BlockDownloader::download_blocks` doesn't necessarily map to one network request? I.e., it can perform multiple requests under the hood, accumulating the data received and returning blocks _as though_ they were downloaded in a single request?

What we are planning to do is to abstract network requests in syncing so that custom syncing protocols ("strategies") can just ask networking to perform a request on their behalf.  The networking/`SyncingEngine` is going to know nothing about the request and treat it only as a pair of protocol name and payload.

There is going to be a way to run syncing strategies in parallel (this is needed to introduce Sync 2.0 not dropping Sync 1.0), but it still serves a different purpose than a custom block downloader you are using, which plugs into standard syncing strategies.

Reply:
> if you are fetching transactions/extrinsics from "mempool", downloading the missing ones, do I understand correctly that BlockDownloader::download_blocks doesn't necessarily map to one network request? I.e., it can perform multiple requests under the hood, accumulating the data received and returning blocks as though they were downloaded in a single request?

Yes, one request can lead to multiple handshake messages between the client/server. The block downloader can be seen as a leaf level/bottom layer helper for fetching a block, using a custom protocol

Reply:
> Yes, one request can lead to multiple handshake messages between the client/server.

Are these handshake messages implemented using standard substrate request-response protocol, or they are a completely custom implementation? I mean on the application layer?

Reply:
we currently use the request-response protocol(with protocol specific endpoint) + SCALE encoding. But this is more of an implementation detail. Some other protocol implementation could choose a different mechanism as it sees best

Reply:
I don't see why `BlockDownloader` or the entire compact block implementation couldn't be expressed as a strategy. `BlockDownloader::download_block()` could be a signal akin to `SyncingService::request_justification()` which could potentially allow us to generalize that system. But that's a future task either way and not in the scope of this PR